lib/ansible/modules/system/gconftool2.py

```diff
@@ -197,74 +285,28 @@ def main():
         supports_check_mode=True
     )
 
-    state_values = {"present": "set", "absent": "unset", "get": "get"}
-
-    direct = False
-    # Assign module values to dictionary values
-    key = module.params['key']
-    value_type = module.params['value_type']
-    if module.params['value'].lower() == "true":
-        value = "true"
-    elif module.params['value'] == "false":
-        value = "false"
-    else:
-        value = module.params['value']
-
-    state = state_values[module.params['state']]
-    if module.params['direct'] in BOOLEANS_TRUE:
-        direct = True
-    config_source = module.params['config_source']
-
-    # Initialize some variables for later
-    change = False
-    new_value = ''
-
-    if state != "get":
-        if value is None or value == "":
-            module.fail_json(msg='State %s requires "value" to be set'
-                             % str(state))
-        elif value_type is None or value_type == "":
-            module.fail_json(msg='State %s requires "value_type" to be set'
-                             % str(state))
-
-        if direct and config_source is None:
-            module.fail_json(msg='If "direct" is "yes" then the ' +
-                             '"config_source" must be specified')
-        elif not direct and config_source is not None:
-            module.fail_json(msg='If the "config_source" is specified ' +
-                             'then "direct" must be "yes"')
-
-    # Create a gconf2 preference
-    gconf_pref = GConf2Preference(module, key, value_type,
-                                  value, direct, config_source)
-    # Now we get the current value, if not found don't fail
-    _, current_value = gconf_pref.call("get", fail_onerr=False)
-
-    # Check if the current value equals the value we want to set.  If not, make
-    # a change
-    if current_value != value:
-        # If check mode, we know a change would have occured.
-        if module.check_mode:
-            # So we will set the change to True
-            change = True
-            # And set the new_value to the value that would have been set
-            new_value = value
-        # If not check mode make the change.
-        else:
-            change, new_value = gconf_pref.call(state)
-    # If the value we want to set is the same as the current_value, we will
-    # set the new_value to the current_value for reporting
-    else:
-        new_value = current_value
-
-    facts = dict(gconftool2={'changed': change,
-                             'key': key,
-                             'value_type': value_type,
-                             'new_value': new_value,
-                             'previous_value': current_value,
-                             'playbook_value': module.params['value']})
-
-    module.exit_json(changed=change, ansible_facts=facts)
+    # If direct access was requested, config_source must be provided.
+    if module.params['direct'] and (module.params['config_source'] is None):
```

Comment:
If config_source is only used in combo with direct, whichs looks to be the case, it may make sense to collapse it into one arg (config_source). Then GConfTool2.__init__ could:

``` python

if config_source:
   self.base_command += ["--direct", "--config-source", config_source]

```

That assumes there won't be a scenario where it makes sense to call 'gconftool-2 --direct' or 'gconftool-2 --config-souce some_config_source' independently.